slasher/db/block_header.go

```diff
@@ -0,0 +1,114 @@
+	var bha []*ethpb.BeaconBlockHeader
+	err := db.view(func(tx *bolt.Tx) error {
+		c := tx.Bucket(historicBlockHeadersBucket).Cursor()
+		prefix := encodeEpochValidatorID(epoch, validatorID)
+		for k, v := c.Seek(prefix); k != nil && bytes.HasPrefix(k, prefix); k, v = c.Next() {
+			bh, err := createBlockHeader(v)
+			if err != nil {
+				return err
+			}
+			bha = append(bha, bh)
+		}
+		return nil
+	})
+	return bha, err
+}
+
+// HasBlockHeader accepts an epoch and validator id and returns true if the block header exists.
+func (db *Store) HasBlockHeader(epoch uint64, validatorID uint64) bool {
+	prefix := encodeEpochValidatorID(epoch, validatorID)
+	var hasBlockHeader bool
+	// #nosec G104
+	_ = db.view(func(tx *bolt.Tx) error {
+		c := tx.Bucket(historicBlockHeadersBucket).Cursor()
+		for k, _ := c.Seek(prefix); k != nil && bytes.HasPrefix(k, prefix); k, _ = c.Next() {
+			hasBlockHeader = true
+			return nil
+		}
+		hasBlockHeader = false
+		return nil
+	})
+
+	return hasBlockHeader
+}
+
+// SaveBlockHeader accepts a block header and writes it to disk.
+func (db *Store) SaveBlockHeader(epoch uint64, validatorID uint64, blockHeader *ethpb.BeaconBlockHeader) error {
+	key := encodeEpochValidatorIDSig(epoch, validatorID, blockHeader.Signature)
+	enc, err := proto.Marshal(blockHeader)
+	if err != nil {
+		return errors.Wrap(err, "failed to encode block")
+	}
+
+	err = db.update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(historicBlockHeadersBucket)
+		if err := bucket.Put(key, enc); err != nil {
+			return errors.Wrap(err, "failed to include the block header in the historic block header bucket")
+		}
+
+		return err
+	})
+
+	// prune history to max size every 10th epoch
+	if epoch%10 == 0 {
+		weakSubjectivityPeriod := uint64(54000)
+		err = db.pruneHistory(epoch, weakSubjectivityPeriod)
+	}
+	return err
+}
+
+// DeleteBlockHeader deletes a block header using the slot and its root as keys in their respective buckets.
+func (db *Store) DeleteBlockHeader(epoch uint64, validatorID uint64, blockHeader *ethpb.BeaconBlockHeader) error {
+
+	key := encodeEpochValidatorIDSig(epoch, validatorID, blockHeader.Signature)
+
+	return db.update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(historicBlockHeadersBucket)
+		if err := bucket.Delete(key); err != nil {
+			return errors.Wrap(err, "failed to delete the block header from historic block header bucket")
+		}
+		return bucket.Delete(key)
+	})
+}
+
+func (db *Store) pruneHistory(currentEpoch uint64, historySize uint64) error {
+	pruneTill := int64(currentEpoch) - int64(historySize)
+	if pruneTill <= 0 {
+		return nil
+	}
+	return db.update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(historicBlockHeadersBucket)
+		c := tx.Bucket(historicBlockHeadersBucket).Cursor()
+		max := bytesutil.Bytes8(uint64(pruneTill))
+		for k, _ := c.First(); k != nil && bytes.Compare(k[:8], max) <= 0; k, _ = c.Next() {
```

Comment:
why do a byte comparison instead of an numerical comparison ?

Reply:
keys are bytes i think it should be faster to keep it like this

Reply:
`bytes.Compare` does a lexicographical comparison , if you wanted to do a numerical one wouldn't it be easier just to use the integer from the key ?

Reply:
+1

Reply:
using byte utils?

Reply:
Yeap, just take the integer out from the first 8 bytes of the key. Then you can simply compare the `uint64` vals directly. Since we are doing a numerical comparison it would be better to compare their vals rather than their byte representations

Reply:
changed